### PR TITLE
Replace FMOD with Kira

### LINKED ISSUE #99: Replace FMOD with a Rust specific 3D Audio Library

### PR DESCRIPTION
**Pull request by HaronK**
Closes #99
[Discord discussion](https://discord.com/channels/695088786702336000/1118643264795447406)

### CODE REVIEW

**Review by IllustrisJack — APPROVED** [review]
Lua looks good to me, wait for @dgavedissian to check the rust part ;)

File: engine/lib/phx/script/ffi/Sound.lua [hunk 1]
```diff
@@ -5,81 +5,39 @@ local Sound
...
-    void   Sound_Rewind                (Sound*);
-    bool   Sound_Get3D                 (Sound*);
-    float  Sound_GetDuration           (Sound*);
-    bool   Sound_GetLooped             (Sound*);
-    cstr   Sound_GetName               (Sound*);
-    cstr   Sound_GetPath               (Sound*);
-    bool   Sound_IsFinished            (Sound*);
-    bool   Sound_IsPlaying             (Sound*);
-    bool   Sound_IsAudible             (Sound*);
-    void   Sound_Attach3DPos           (Sound*, Vec3f const* pos, Vec3f const* vel);
```
Comment (dgavedissian) [inline engine/lib/phx/script/ffi/Sound.lua:25]: We removed quite a lot of functions here, are we certain we don't need them?

File: engine/lib/phx/src/audio/audio.rs [hunk 2]
```diff
@@ -1,227 +1,83 @@
...
-pub unsafe extern "C" fn Audio_AllocSound() -> *mut Sound {
-    MemPool_Alloc(&mut *this.soundPool) as *mut Sound
-}
-
-#[no_mangle]
-pub unsafe extern "C" fn Audio_DeallocSound(sound: *mut Sound) {
-    MemPool_Dealloc(&mut *this.soundPool, sound as *mut _);
-}
+    pub fn get_playing_count(&self) -> u64 {
+        0 // FIXME
```
Comment (dgavedissian) [inline engine/lib/phx/src/audio/audio.rs:77]: Was this missed, or being addressed later?

File: engine/lib/phx/src/audio/sound.rs [hunk 3]
```diff
@@ -1,500 +1,164 @@
...
-                    ) -> FMOD_RESULT,
-            ),
-        ));
-        Sound_SetState(this, 2 as SoundState);
-        if Sound_Get3D(this) {
-            let mut zero: Vec3 = Vec3::ZERO;
-            Sound_Set3DPos(this, &mut zero, &mut zero);
-        }
+    pub fn set_emitter(&mut self, emitter: EmitterHandle) {
+        // FIXME: for some reason adding emitter kills the sound
```
Comment (dgavedissian) [inline engine/lib/phx/src/audio/sound.rs:35]: Does this mean this function doesnt work?

**Comment by dgavedissian** [comment]
I noticed that AudioTest doesn't work currently:
```
λ cargo r -- AudioTest
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /Users/dgavedissian/Work/ltheory/engine/lib/phx/Cargo.toml
workspace: /Users/dgavedissian/Work/ltheory/Cargo.toml
    Finished dev [unoptimized + debuginfo] target(s) in 0.31s
     Running `target/debug/ltr AudioTest`
Engine_Init: Requesting GL 2.1
thread '<unnamed>' panicked at 'Cannot load sound: IoError(Os { code: 2, kind: NotFound, message: "No such file or directory" })', engine/lib/phx/src/audio/sound.rs:59:69
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
fatal runtime error: failed to initiate panic, error 5
Signal_Handler: Caught Abort
```

Can we avoid panicking if the sound doesnt exist, or at least, log the resource name it's trying to load?

File: engine/lib/phx/script/ffi/Sound.lua (hunk 1 above)
Comment (HaronK) [reply engine/lib/phx/script/ffi/Sound.lua:25]: We don't use them, that's for sure. We can add them later if needed.

File: engine/lib/phx/src/audio/audio.rs (hunk 2 above)
Comment (HaronK) [reply engine/lib/phx/src/audio/audio.rs:77]: Honestly I didn't find where to take this number from yet. :)

File: engine/lib/phx/src/audio/audio.rs [hunk 4]
```diff
@@ -1,227 +1,83 @@
...
     }
-    for i in soundsToRemove.iter().rev() {
-        this.playingSounds.swap_remove(*i);
-    }
 
-    for sound in this.freeingSounds.iter() {
-        Audio_DeallocSound(*sound);
+    pub fn set_listener_pos(&mut self, pos: &Vec3, rot: &Quat) {
+        self.listener
+            .set_position(*pos, Tween::default()) // TODO: set correct tween
```
[the remark this answers is not in the document]
Comment (HaronK) [reply engine/lib/phx/src/audio/audio.rs:65]: I think we can use default for now. If there will be a problem then we can fix it.

File: engine/lib/phx/src/audio/sound.rs (hunk 3 above)
Comment (HaronK) [reply engine/lib/phx/src/audio/sound.rs:35]: Yes, setting emitter doesn't work but without it sound plays ok, so I'll postpone investigation of this issue for later.

File: engine/lib/phx/src/audio/audio.rs (hunk 2 above)
Comment (HaronK) [reply engine/lib/phx/src/audio/audio.rs:77]: I found a potential way to implement this but it is not trivial. Each Sound object has method is_playing() so we can go over all sounds and find who is playing. But at the moment we don't have common register of sounds.

So we can keep all sounds in Audio object. In this case we will have to use Rc/RefCell (or even Arc/Mutex) to share sounds between Rust and Lua code.

I think for now it's too complicated to implement. We can take a look at it later.

**Comment by HaronK** [comment]
Added a file name to see which sound is missing when `cargo r -- AudioTest` command fails.

**Pull request merged**